File: app/src/main/java/netpro/keyTransmitter/PressingKey.java
package netpro.keytransmitter;

/**
 * 押している間キーコードを送信し続けるキーです
 */
public class PressingKey extends Key  {

    private boolean running;
    private long inputInterval;

    /**
     * @param inputInterval キーを送信する間隔(ミリ秒)
     */
    public PressingKey(int columnSpan, int rowSpan,String description, Type type, long inputInterval) {
        super(columnSpan, rowSpan, description, type);
        this.inputInterval = inputInterval;
    }

    @Override
    public void onActionDown() {
        running = false;
        running = true;
        new Thread(new Executer()).start();
    }

    @Override
    public void onActionUp() {
        running = false;
    }

    @Override
    public void onCancel() {
        running = false;
    }

    public long getInputInterval() {
        return inputInterval;
    }

    private class Executer implements Runnable {
        @Override
        public void run() {
            while (running) {
                send();
                try {
                    Thread.sleep(inputInterval);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }
        }
    }
}

File: app/src/main/java/netpro/keytransmitter/PressingKey.kt
package netpro.keytransmitter

import android.view.MotionEvent
import android.view.View

/**
 * 押している間、設定した間隔でキーコードを送信し続けるボタンです
 */
class PressingKey
/**
 * @param inputInterval キーを送信する間隔(ミリ秒)
 */
(columnSpan: Int, rowSpan: Int, description: String, type: KeyType, val inputInterval: Long) : BaseKey(columnSpan, rowSpan, description, type) {

    val NORMAL = "normal"

    private var running: Boolean = false

    override fun onActionDown(view: View, motionEvent: MotionEvent) {
        running = false
        running = true
        Thread(Executor()).start()
    }

    override fun onActionUp(view: View, motionEvent: MotionEvent) {
        running = false
    }

    override fun onCancel() {
        running = false
    }

    private inner class Executor : Runnable {
        override fun run() {
            while (running) {
                send(keyCodesMap[NORMAL])
                try {
                    Thread.sleep(inputInterval)
                } catch (e: InterruptedException) {
                    e.printStackTrace()
                }

            }
        }
    }

    companion object {
        private val serialVersionUID = 7854266457351207301L
    }
}
